lib/fs.js

```diff
@@ -2009,15 +2009,31 @@ function ReadStream(path, options) {
   this.closed = false;
 
   if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
+    // Case 1: If the type of start is not 'number', throw ERR_INVALID_ARG_TYPE
+    // (TypeError).
+    if (typeof this.start !== 'number') {
       throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
     }
-    if (this.end === undefined) {
-      this.end = Infinity;
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+
+    // Case 2: If start is NaN, throw ERR_OUT_OF_RANGE (RangeError).
+    if (Number.isNaN(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'not NaN', this.start);
+    }
+
+    // Case 3: If start is negative, throw ERR_OUT_OF_RANGE (RangeError).
+    if (this.start < 0) {
+      throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+    }
+
+    // Case 4: If start is fractional, round them to the nearest whole number.
+    // (Not integer as negatives not allowed)
+    if (!Number.isInteger(this.start)) {
+      this.start = Math.round(this.start);
     }
 
+    // Case 5: If start is a whole number, work perfectly.
+
+    // Case 6: If start is greater than end, throw ERR_OUT_OF_RANGE (RangeError)
     if (this.start > this.end) {
```

Comment:
I think this works fine, but as far as I can tell that is because `end: undefined` gets coerced to `NaN` here (which makes this condition always false), and that the fact that we later round `end` does not affect this comparison

Maybe it would be a bit more obvious that this works if we swap type checking for `end` and `start`?

Reply:
@ryzokuken What I meant was, `end` is being coerced to `NaN` in this comparison if it was passed in as `undefined`, because we haven’t yet gotten to the part where it’s converted to `Infinity`.

Reply:
Oh. so, should I move the block that handles `this.end === undefined` first and move to a similar type-check below?

Let me make it and demonstrate.